include/swift/AST/Expr.h

```diff
@@ -4593,22 +4593,27 @@ class DotSyntaxBaseIgnoredExpr : public Expr {
 class ExplicitCastExpr : public Expr {
   Expr *SubExpr;
   SourceLoc AsLoc;
-  TypeExpr *const CastTy;
+  Type CastTy;
+  TypeRepr *const CastRepr;
 
 protected:
-  ExplicitCastExpr(ExprKind kind, Expr *sub, SourceLoc AsLoc, TypeExpr *castTy)
+  ExplicitCastExpr(ExprKind kind, Expr *sub, SourceLoc AsLoc, TypeRepr *repr)
```

Comment:
Does it make sense to have an overload of constructor here which accepts `Type castTy` as a last argument instead of having to use `setCastType` every time there is no type repr?

Reply:
Maybe it could be a `PointerUnion<TypeRepr *, TypeBase *>` actually to avoid adding a second overload to each cast expression.

Reply:
Now we're just rebuilding `TypeLoc`...

Reply:
Are there any other places where we need to do this?

Reply:
@xedin – Ya, I toyed with having an overloaded constructor. We still can. I was just trying to keep the patch small to ease review. I don't think using a `PointerUnion` adds value over simply overloading the constructor.

@CodaFi – I support the goal of not reinventing `TypeLoc` and towards that end, that's why I made `setCastType` to assert that people are NOT changing the type, only setting it.